{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "# Function to extract details from the FASTA header\n",
    "def extract_details(description):\n",
    "    # Extract id (e.g., ENSMUST00000103301.3)\n",
    "    id_match = re.search(r\"(ENST\\d+\\.\\d+)\", description)\n",
    "    id_ = id_match.group(1) if id_match else None\n",
    "\n",
    "    # Extract gene (e.g., ENSMUSG00000076500.3)\n",
    "    gene_match = re.search(r\"gene:(ENSG\\d+\\.\\d+)\", description)\n",
    "    gene = gene_match.group(1) if gene_match else None\n",
    "\n",
    "    # Example: gene_symbol:Gm10878\n",
    "    gene_symbol_match = re.search(r\"gene_symbol:([^\\s]+)\", description)\n",
    "    gene_symbol = gene_symbol_match.group(1) if gene_symbol_match else None\n",
    "\n",
    "    # Extract chromosome (e.g., GRCh38:6)\n",
    "    chromosome_match = re.search(r\"chromosome:(GRCh38:\\d+)\", description)\n",
    "    chromosome = chromosome_match.group(1) if chromosome_match else None\n",
    "\n",
    "    # Extract start, end, and strand\n",
    "    position_match = re.search(r\"chromosome:GRCh38:(\\d+):(\\d+):(\\d+):(-?1)\", description)\n",
    "    start = int(position_match.group(2)) if position_match else None\n",
    "    end = int(position_match.group(3)) if position_match else None\n",
    "    strand = int(position_match.group(4)) if position_match else None\n",
    "\n",
    "    return id_, gene, gene_symbol, chromosome, start, end, strand\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Function to parse the FASTA file and store in DataFrame\n",
    "def parse_fasta_to_df(fasta_file):\n",
    "    # List to hold all entries\n",
    "    records = []\n",
    "\n",
    "    # Parse the FASTA file/\n",
    "    for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        # Extract id, gene, chromosome, start, end, and strand\n",
    "        id_, gene, gene_symbol, chromosome, start, end, strand = extract_details(record.description)\n",
    "        \n",
    "        # Create a dictionary for the current entry\n",
    "        entry = {\n",
    "            \"id\": id_,\n",
    "            \"gene\": gene,\n",
    "            \"gene_symbol\": gene_symbol,\n",
    "            \"chromosome\": chromosome,\n",
    "            \"start\": start,\n",
    "            \"end\": end,\n",
    "            \"strand\": strand,\n",
    "            #\"Description\": record.description,\n",
    "            \"nuc_seq\": str(record.seq)\n",
    "        }\n",
    "        # Append to the list\n",
    "        records.append(entry)\n",
    "    \n",
    "    # Convert the list of dicts to DataFrame\n",
    "    df = pd.DataFrame(records)\n",
    "    return df\n",
    "\n",
    "# Usage example with the sample file\n",
    "df = parse_fasta_to_df(\"./data/Homo_sapiens.GRCh38.cdna.all.fa\")\n",
    "df[\"ens_gene_id\"] = df[\"gene\"].str.split(\".\").str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>nuc_seq</th>\n",
       "      <th>ens_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000390473.1</td>\n",
       "      <td>ENSG00000211825.1</td>\n",
       "      <td>TRDJ1</td>\n",
       "      <td>GRCh38:14</td>\n",
       "      <td>22450089.0</td>\n",
       "      <td>22450139.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ACACCGATAAACTCATCTTTGGAAAAGGAACCCGTGTGACTGTGGA...</td>\n",
       "      <td>ENSG00000211825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000390484.1</td>\n",
       "      <td>ENSG00000211836.1</td>\n",
       "      <td>TRAJ54</td>\n",
       "      <td>GRCh38:14</td>\n",
       "      <td>22482287.0</td>\n",
       "      <td>22482346.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TAATTCAGGGAGCCCAGAAGCTGGTATTTGGCCAAGGAACCAGGCT...</td>\n",
       "      <td>ENSG00000211836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000390488.1</td>\n",
       "      <td>ENSG00000211840.1</td>\n",
       "      <td>TRAJ49</td>\n",
       "      <td>GRCh38:14</td>\n",
       "      <td>22489488.0</td>\n",
       "      <td>22489543.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GAACACCGGTAACCAGTTCTATTTTGGGACAGGGACAAGTTTGACG...</td>\n",
       "      <td>ENSG00000211840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000390476.1</td>\n",
       "      <td>ENSG00000211828.1</td>\n",
       "      <td>TRDJ3</td>\n",
       "      <td>GRCh38:14</td>\n",
       "      <td>22459098.0</td>\n",
       "      <td>22459156.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CTCCTGGGACACCCGACAGATGTTTTTCGGAACTGGCATCAAACTC...</td>\n",
       "      <td>ENSG00000211828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000390489.1</td>\n",
       "      <td>ENSG00000211841.1</td>\n",
       "      <td>TRAJ48</td>\n",
       "      <td>GRCh38:14</td>\n",
       "      <td>22490491.0</td>\n",
       "      <td>22490553.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TATCTAACTTTGGAAATGAGAAATTAACCTTTGGGACTGGAACAAG...</td>\n",
       "      <td>ENSG00000211841</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id               gene gene_symbol chromosome       start  \\\n",
       "0  ENST00000390473.1  ENSG00000211825.1       TRDJ1  GRCh38:14  22450089.0   \n",
       "1  ENST00000390484.1  ENSG00000211836.1      TRAJ54  GRCh38:14  22482287.0   \n",
       "2  ENST00000390488.1  ENSG00000211840.1      TRAJ49  GRCh38:14  22489488.0   \n",
       "3  ENST00000390476.1  ENSG00000211828.1       TRDJ3  GRCh38:14  22459098.0   \n",
       "4  ENST00000390489.1  ENSG00000211841.1      TRAJ48  GRCh38:14  22490491.0   \n",
       "\n",
       "          end  strand                                            nuc_seq  \\\n",
       "0  22450139.0     1.0  ACACCGATAAACTCATCTTTGGAAAAGGAACCCGTGTGACTGTGGA...   \n",
       "1  22482346.0     1.0  TAATTCAGGGAGCCCAGAAGCTGGTATTTGGCCAAGGAACCAGGCT...   \n",
       "2  22489543.0     1.0  GAACACCGGTAACCAGTTCTATTTTGGGACAGGGACAAGTTTGACG...   \n",
       "3  22459156.0     1.0  CTCCTGGGACACCCGACAGATGTTTTTCGGAACTGGCATCAAACTC...   \n",
       "4  22490553.0     1.0  TATCTAACTTTGGAAATGAGAAATTAACCTTTGGGACTGGAACAAG...   \n",
       "\n",
       "       ens_gene_id  \n",
       "0  ENSG00000211825  \n",
       "1  ENSG00000211836  \n",
       "2  ENSG00000211840  \n",
       "3  ENSG00000211828  \n",
       "4  ENSG00000211841  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>nuc_seq</th>\n",
       "      <th>ens_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000390473.1</td>\n",
       "      <td>ENSG00000211825.1</td>\n",
       "      <td>TRDJ1</td>\n",
       "      <td>GRCh38:14</td>\n",
       "      <td>22450089.0</td>\n",
       "      <td>22450139.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ACACCGATAAACTCATCTTTGGAAAAGGAACCCGTGTGACTGTGGA...</td>\n",
       "      <td>ENSG00000211825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000390484.1</td>\n",
       "      <td>ENSG00000211836.1</td>\n",
       "      <td>TRAJ54</td>\n",
       "      <td>GRCh38:14</td>\n",
       "      <td>22482287.0</td>\n",
       "      <td>22482346.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TAATTCAGGGAGCCCAGAAGCTGGTATTTGGCCAAGGAACCAGGCT...</td>\n",
       "      <td>ENSG00000211836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000390488.1</td>\n",
       "      <td>ENSG00000211840.1</td>\n",
       "      <td>TRAJ49</td>\n",
       "      <td>GRCh38:14</td>\n",
       "      <td>22489488.0</td>\n",
       "      <td>22489543.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GAACACCGGTAACCAGTTCTATTTTGGGACAGGGACAAGTTTGACG...</td>\n",
       "      <td>ENSG00000211840</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id               gene gene_symbol chromosome       start  \\\n",
       "0  ENST00000390473.1  ENSG00000211825.1       TRDJ1  GRCh38:14  22450089.0   \n",
       "1  ENST00000390484.1  ENSG00000211836.1      TRAJ54  GRCh38:14  22482287.0   \n",
       "2  ENST00000390488.1  ENSG00000211840.1      TRAJ49  GRCh38:14  22489488.0   \n",
       "\n",
       "          end  strand                                            nuc_seq  \\\n",
       "0  22450139.0     1.0  ACACCGATAAACTCATCTTTGGAAAAGGAACCCGTGTGACTGTGGA...   \n",
       "1  22482346.0     1.0  TAATTCAGGGAGCCCAGAAGCTGGTATTTGGCCAAGGAACCAGGCT...   \n",
       "2  22489543.0     1.0  GAACACCGGTAACCAGTTCTATTTTGGGACAGGGACAAGTTTGACG...   \n",
       "\n",
       "       ens_gene_id  \n",
       "0  ENSG00000211825  \n",
       "1  ENSG00000211836  \n",
       "2  ENSG00000211840  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.to_csv(\"./data/nuc_seqs_human.csv\", index=False)\n",
    "display(df.head(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "18:10:12 - INFO - Requesting amino acid sequence of the canonical transcript ENST00000390473 of gene ENSG00000211825 from UniProt.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['>ENST00000390473 uniprot_id: A0A075B706 ensembl_id: ENST00000390473 gene_name: TRDJ1 organism: Homo sapiens sequence_length: 16', 'TDKLIFGKGTRVTVEP']\n"
     ]
    }
   ],
   "source": [
    "import gget\n",
    "\n",
    "out = gget.seq(df[\"ens_gene_id\"][0], translate=True, isoforms=False)\n",
    "print(out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def parse_seq_header(header):\n",
    "    # Extract key information using regex\n",
    "    ensemble_pattern = r'ensembl_id:\\s*(ENS[T|G]\\d+)'\n",
    "    gene_pattern = r'gene_name:\\s*([^\\s]+)'  # Changed: capture everything until a space\n",
    "    length_pattern = r'sequence_length:\\s*(\\d+)'\n",
    "    \n",
    "    ensemble_id = re.search(ensemble_pattern, header).group(1)\n",
    "    gene_name = re.search(gene_pattern, header).group(1)\n",
    "    sequence_length = int(re.search(length_pattern, header).group(1))\n",
    "    \n",
    "    return {\n",
    "        'ensemble_id': ensemble_id,\n",
    "        'gene_name': gene_name,\n",
    "        'sequence_length': sequence_length\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "18:10:20 - INFO - Requesting amino acid sequence of the canonical transcript ENST00000390473 of gene ENSG00000211825 from UniProt.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensemble_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>sequence_length</th>\n",
       "      <th>aa_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000390473</td>\n",
       "      <td>TRDJ1</td>\n",
       "      <td>16</td>\n",
       "      <td>TDKLIFGKGTRVTVEP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensemble_id gene_name  sequence_length            aa_seq\n",
       "0  ENST00000390473     TRDJ1               16  TDKLIFGKGTRVTVEP"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out = gget.seq(df[\"ens_gene_id\"][0], translate=True, isoforms=False)\n",
    "if len(out)==2:\n",
    "    df_prot = pd.DataFrame([parse_seq_header(out[0])])\n",
    "    df_prot[\"aa_seq\"] = out[1]\n",
    "df_prot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
